transactions/read-only/parallel.md

```diff
@@ -0,0 +1,215 @@
+
+## Design Decisions
+
+### Window Toggling
+To support toggling between `read` and `write` windows, configurable options `read-window-time`, `write-window-time`, and `read-only-max-queued-time` are provided. 
+- From `write` to `read`: When a new read-only transaction is queued, if the time the earliest transaction has been queued exceeds `read-only-max-queued-time`; or at the end of the `write` window and read-only transaction queue not empty.
+- From `read` to `write`: when read-only transaction queue is emptied by the read-only threads; or at the end of `read` window.
+
+In the following state diagram, `longest_queued_time = now - the time when the oldest trx was queued`, `write_window_deadline = time when write window starts + write-window-time`, and `read_window_deadline = time when read window starts + read-window-time`
+
+```mermaid
+flowchart TD
+    A(((write window))) -->|push a new trx| B[longest_queued_time < read-only-max-queued-time?]
+    B -->|yes| R(((read window)))
+    B -->|no| A
+    A --> D[write_window_deadline passed?]
+    D -->|yes| F[read-only trx queue empty?]
+    D -->|no| A
+    F -->|yes| A
+    F -->|no| R
+    
+    R --> S[read-only trx queue empty?]
+    S -->|yes| A
+    S -->|no| R
+    R --> T[read_window_deadline passed?]
+    T -->|yes| A
+    T -->|no| R  
+```
+
+### Handling main thread functions that are not safe to read-only transaction execution in `read` Window
+Several options are considered to handle 
+- Drop the functions. This in not acceptable as it changes the behavior.
+- Re-post to the main thread. In `read` window, re-post a not-read-only-safe function back to the main thread. This is simple to implement but breaks the order of functions and waste time in popping and pushing. 
+- Modify appbase `execution_priority_queue` by adding a function type with value `read-only-safe` or `not-read-only-safe`. In `write` window, everything works as it is now. In `read` window, only `thread-safe` requests are dequeued and processed. This avoids introducing new queues and keeps the order of the request. But as the queue is a priority queue based on function priority, it is infeasible to incorporate the additional type into the priority in a single queue.
+- Separate appbase queue into two queues: `read-only-safe` or `not-read-only-safe`. The function type is added when a function is posted. 
+  - In `read` window, only functions in `read-only-safe` queue are executed.
+  - In `write` window, compare the priorities of the top functions in `read-only-safe` and `not-read-only-safe`. The one with higher priority is executed. If tied, three options are considered:
+    - `not-read-only-safe` function is favored 
+    - randomly pick one
+    - add a time attribute to the functions and the older one is picked. This keeps the original behavior. Even though at a cost of the extra time field and an extra comparison, this option seems best.
```

Comment:
Would be quicker to just use an `uint64_t` insert counter instead of time so you don't have to access any clock.